### PR TITLE
global: elasticsearch percolator

### PR DESCRIPTION
**Pull request by hachreak**
Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio_collections/models.py [hunk 1]
```diff
@@ -58,4 +59,70 @@ def collection_attribute_changed(target, value, oldvalue, initiator):
 listen(Collection.dbquery, 'set', collection_attribute_changed)
 listen(Collection.parent_id, 'set', collection_attribute_changed)
 
+# percolator
+
+
+@celery.task
```
Comment (jirikuncar) [inline invenio_collections/models.py]: use `shared_task` instead

File: invenio_collections/models.py [hunk 2]
```diff
@@ -58,4 +59,70 @@ def collection_attribute_changed(target, value, oldvalue, initiator):
...
+    """Delete percolator when collection is deleted."""
+    if target.dbquery is not None:
+        delete_collection_percolator.delay(target)
+
+
+listen(Collection, 'after_insert', collection_inserted_percolator)
+listen(Collection, 'after_delete', collection_removed_percolator)
+listen(Collection.dbquery, 'set', collection_attribute_changed_percolator)
+# FIXME can I update percolator id without recreate?
+listen(Collection.name, 'set', collection_name_changed_percolator)
```
Comment (jirikuncar) [inline invenio_collections/models.py]: probably it is better to drop the old one and create new one

**Comment by hachreak** [comment]
ping @jirikuncar

File: invenio_collections/receivers.py [hunk 3]
```diff
@@ -58,7 +58,39 @@ def _build_cache():
...
+    """Find matching collections with percolator engine.
+
+    :param collections: set of collections where search
+    :param record: record to match
+    """
+    from invenio_search import current_search_client
+    results = current_search_client.percolate(
+        index='records',
+        doc_type='record',
+        doci=record.json,
```
Comment (jirikuncar) [inline invenio_collections/receivers.py]: `doci` vs `doc`?

**Comment by jirikuncar** [comment]
@hachreak it looks good. Please include more tests. Thanks

File: tests/conftest.py [hunk 4]
```diff
@@ -65,3 +69,83 @@ def teardown():
         db.create_all()
 
     return app
+
+
+@pytest.fixture()
+def esapp(request):
+    """Create Flask app fixture for Elastic Search."""
+    # Create Flask application
+    app = Flask(__name__)
```
Comment (jirikuncar) [inline tests/conftest.py]: you can reuse the `app` fixture

File: invenio_collections/config.py [hunk 5]
```diff
@@ -38,11 +38,17 @@
 COLLECTIONS_REGISTER_RECORD_SIGNALS = True
 """Catch record insert/update signals and update the `_collections` field."""
 
-try:
-    pkg_resources.get_distribution('invenio_search')
+# try:
```
Comment (jirikuncar) [inline invenio_collections/config.py]: you can uncomment it

**Comment by hachreak** [comment]
ping @jirikuncar

File: invenio_collections/percolator.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    from elasticsearch import NotFoundError
+    for name in current_search.mappings.keys():
+        if target.name and target.dbquery:
+            try:
+                current_search.client.delete(
+                    index=name,
+                    doc_type='.percolator',
+                    id='collection-{}'.format(target.name),
+                )
+            except NotFoundError:
```
Comment (jirikuncar) [inline invenio_collections/percolator.py]: you can use `ignore=[404]` in client delete

File: invenio_collections/ext.py [hunk 7]
```diff
@@ -26,9 +26,15 @@
...
 from __future__ import absolute_import, print_function
 
+from invenio_records import signals
+from sqlalchemy.event import contains, listen, remove
+
 from . import config
 from .cli import collections as collections_cmd
-from .receivers import update_collections
+from .models import Collection
+from .percolator import collection_inserted_percolator, \
```
Comment (jirikuncar) [inline invenio_collections/ext.py]: move the import to line 70

File: invenio_collections/percolator.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Percolator."""
+
+
+def new_collection_percolator(target):
+    """Create new percolator associated with the new collection."""
+    from invenio_search import current_search
```
Comment (jirikuncar) [inline invenio_collections/percolator.py]: these imports can be global then

File: invenio_collections/receivers.py [hunk 9]
```diff
@@ -56,9 +56,49 @@ def _build_cache():
...
+    :param collections: set of collections where search
+    :param record: record to match
+    """
+    for name, data in iteritems(collections):
+        if _build_query(data['query']).match(record):
+            yield data['ancestors']
+    raise StopIteration
+
+
+def _find_matching_collections_externally(collections, record):
```
Comment (jirikuncar) [inline invenio_collections/receivers.py]: you can move this function to `percolator.py`

File: invenio_collections/receivers.py [hunk 10]
```diff
@@ -66,17 +106,34 @@ def get_record_collections(record):
...
-            output.add(name)
-            output |= data['ancestors']
+    for collections in matcher(collections, record):
+        output |= collections
+
     return list(output)
 
 
-def update_collections(sender, *args, **kwargs):
+def update_collections(sender, matcher, *args, **kwargs):
```
Comment (jirikuncar) [inline invenio_collections/receivers.py]: it should be inside `__call__` method

File: invenio_collections/ext.py [hunk 11]
```diff
@@ -26,9 +26,13 @@
 
 from __future__ import absolute_import, print_function
 
+from invenio_records import signals
+from sqlalchemy.event import contains, listen, remove
+
 from . import config
 from .cli import collections as collections_cmd
-from .receivers import update_collections
+from .models import Collection
```
Comment (jirikuncar) [inline invenio_collections/ext.py]: do not import models on global level

**Pull request merged**